design-proposals/persist-vmi-firmware-uuid.md

```diff
@@ -0,0 +1,181 @@
+   - Could lead to cluttered and harder-to-read VM definitions.
+
+2. **Generate UUID via Webhook and Store in the Spec Field**  
+   **Description:** If no UUID is specified by the user, a webhook generates a random UUID and assigns it to the VM’s spec field.
+
+   **Pros:**
+   - Similar to the first option, avoiding a new API field and maintaining backward compatibility.
+
+   **Cons:**
+   - Adds complexity by involving webhooks, which could slow down performance if many VMs are started simultaneously.
+   - May potentially bottleneck `virt-api`.
+
+3. **Add a Firmware UUID Field to VM Status**  
+   **Description:** Store the generated UUID within the VM’s status.
+
+   **Pros:**
+   - Keeps the spec clean.
+
+   **Cons:**
+   - Could introduce a new API field, impacting long-term maintenance and adding potential API bloat.
+
+### Phase 2: Transition to Universally Unique UUID Generation
+
+In Phase 2, the approach for generating the firmware UUID is modified to ensure global uniqueness.
+This transition could involve incorporating the VM name and namespace into the UUID,
+or adopting a pseudo-random hash to guarantee uniqueness across clusters.
+The solutions for this phase are outlined below:
+
+1. **Use Namespace + Name for UUID Generation**  
+   **Description:** Generate the UUID by combining the VM name and namespace, ensuring uniqueness within the cluster.
+
+   **Pros:**
+   - Simple and straightforward to implement.
+   - Avoids reliance on an external generator for randomness.
+
+   **Cons:**
+   - Introduces a breaking change, as UUIDs for existing VMs may change unless preserved.
+   - Does not guarantee universal uniqueness across clusters.
+
+2. **Adopt Pseudo-Random Hashing for UUID Generation**  
+   **Description:** Use a pseudo-random hash to generate a unique UUID that does not depend on specific names or namespaces.
+
+   **Pros:**
+   - Ensures global uniqueness, even across clusters.
+   - Provides flexibility by avoiding dependence on VM names and namespaces.
+
+   **Cons:**
+   - Introducing randomness may make the UUID less predictable, potentially impacting certain workloads.
+   - Requires careful validation to ensure the hash remains consistent for each VM across restarts.
+
+By addressing each phase separately, we can balance immediate workload stability
+with a gradual transition to globally unique UUIDs, ensuring minimal disruption for end-users
+while providing a scalable solution for UUID uniqueness.
+
+
+## API Examples
+**Example uuid persistence within VM Status**
+
+```yaml
+apiVersion: kubevirt.io/v1alpha3
+kind: VirtualMachine
+metadata:
+  name: mytestvm
+status:
+   conditions:
+      - lastProbeTime: "2024-11-06T01:12:29Z"
+        lastTransitionTime: "2024-11-06T01:12:29Z"
+        message: Guest VM is not reported as running
+        reason: GuestNotRunning
+        status: "False"
+        type: Ready
+   created: true
+   runStrategy: Once
+   firmwareUUID: "123e4567-e89b-12d3-a456-426614174000"
```

Comment:
.status ins only for erporting, not for persitence.

Reply:
Can you elaborate?
How is spec more persistent than status? Aren't they both persisted in etcd?

Reply:
> .status ins only for erporting, not for persitence.

@fabiand I don't know where this is coming from. Numerous objects across k8s use the status exactly for that purpose, including objects in KubeVirt.

Reply:
The VM object, until it is deleted would have a status - it simply stores the current state of the VM, while the spec holds the desired.

Reply:
Conceptually, it is not what status is meant to show.

Status is meant to represent the current state. However, the FW UUID cannot be "read" from the current state of the VM, because it is stopped.
There is a name to persisting such information on the status: caching.
Status is not about caching, it is about representing the actual. But here the suggestion is to store the actual so it will be the input for a future actual (which comes in the opposite direction).

We actually had this implemented in the network interfaces status and it caused a mess: To calculate the new status on a reconcile cycle, the previous data was considered, potentially dragging incorrect state over cycles indefinitely. We have since dropped this behavior, calculating the interfaces status without depending on the previous state.
But this is still tricky business even today, because multiple components attempt to touch the same list.

Bottom line, while this could be a good solution, caching the data on the status can potentially bite back. There is logic that uses this status information to set the UUID on the next VM instantiation and an opposite logic that reads the VM instantiation spec (or status?) to reflect that actual back to the same status field. Sounds like a loop to me.

Reply:
Even if the VM is stopped the status shows the `current` state at this UUID is stored on the VM boot disk
As long as the VM exists this is its current state, regardless of its phase.